Add basic Embassy support

Serhii-the-Dev opened the pull request:
Resolves #49 

Embassy time driver implementation is mostly a copy-pasta of [the STM32 driver](https://github.com/embassy-rs/embassy/blob/1cf778904d597a5bc01a4b7862f965681636faf1/embassy-stm32/src/time_driver.rs#L275) except it's not implementing RTC functionality required for low-power/deep-sleep scenarios in order to keep it simple (AFAIK current HAL implementation also lacks of RTC support). 

This PR brings a set of new features:
- `embassy` - enables Embassy support in general
- `time-driver-tim<1, 2, 14>` - enables Embassy time driver support for a specified general-purpose timer instance

Both features are required to use the Embassy executor, so any of the `time-driver-tim<1, 2, 14>` features enables the `embassy` feature in general.

Despite that the driver uses `unsafe` access to the selected timer's registry, in order to guard an access to a `TIMERx` peripheral, the driver initialization requires to move a `timer<1, 2, 14>` so it couldn't be used anywhere else causing undefined behavior, ex:
```rust
embassy::init(p.timer1, &clocks, &mut rcu.apb1);
```
After that the `p.timer1` could not be moved, to create a `Timer::timer1`, for example.

Please take into account that Embassy support comes with a price, as a binary relying on the `embassy` feature and using the `async` executor would have additional ~2KiB added to it's size (also an extra memory would be required for the tasks arena):
```bash
> cargo bloat --release --features gd32f130x8,time-driver-tim1 --example embassy
...
File  .text   Size            Crate Name
0.4%  41.9% 1.4KiB embassy_executor embassy_executor::arch::thread::Executor::run
0.2%  20.2%   706B embassy_executor embassy_executor::raw::TaskStorage<F>::poll
0.1%  11.6%   406B     gd32f1x0_hal TIMER1
0.1%  10.5%   366B embassy_executor embassy_executor::raw::TaskStorage<F>::poll
0.0%   4.7%   164B        [Unknown] __aeabi_memclr8
0.0%   2.3%    80B embassy_executor embassy_executor::arch::thread::Executor::new
0.0%   1.3%    46B embassy_executor embassy_executor::raw::util::UninitCell<T>::write_in_place
0.0%   1.1%    40B      cortex_m_rt Reset
0.0%   0.7%    24B        [Unknown] HardFaultTrampoline
0.0%   0.6%    20B          embassy embassy::__cortex_m_rt_main
0.0%   0.5%    18B              std core::option::unwrap_failed
0.0%   0.5%    16B     gd32f1x0_hal gd32f1x0_hal::gpio::gpioc::<impl gd32f1x0_hal::gpio::GpioRegExt for gd32f1::gd32f130::gpioc::RegisterBlock>::is_set_low
0.0%   0.5%    16B     gd32f1x0_hal gd32f1x0_hal::gpio::gpioc::<impl gd32f1x0_hal::gpio::GpioRegExt for gd32f1::gd32f130::gpioc::RegisterBlock>::is_low
0.0%   0.4%    14B     gd32f1x0_hal gd32f1x0_hal::gpio::gpioc::<impl gd32f1x0_hal::gpio::GpioRegExt for gd32f1::gd32f130::gpioc::RegisterBlock>::set_high
0.0%   0.4%    14B     gd32f1x0_hal gd32f1x0_hal::gpio::gpioc::<impl gd32f1x0_hal::gpio::GpioRegExt for gd32f1::gd32f130::gpioc::RegisterBlock>::set_low
0.0%   0.3%    10B         cortex_m _critical_section_1_0_release
0.0%   0.2%     8B embassy_executor embassy_executor::raw::SyncExecutor::alarm_callback
0.0%   0.2%     8B              std core::panicking::panic_const::panic_const_async_fn_resumed
0.0%   0.2%     8B              std core::panicking::panic
0.0%   0.2%     8B              std core::panicking::panic_fmt
0.0%   1.1%    38B                  And 9 smaller methods. Use -n N to show more.
0.8% 100.0% 3.4KiB                  .text section size, the file size is 405.7KiB
```

Serhii-the-Dev committed:
Add basic Embassy support

qwandor committed:
Avoid casting function pointer.

Serhii-the-Dev commented:
@qwandor I must ask to wait with a merge as I've just discovered some strange issues with this PR: a couple of fresh new GD32F130C8 are starting to reboot after a few alarm cycles on any timer/channel instance. I tried to fully erase my GD32F130G8's flash, and uploaded an example, now it's reproducing on this chip as well. Moreover, openocd is not able to program the GD32F130G8 anymore, however probe-rs is working just fine. Alas, no ideas what's going so far, I'll keep investigation.

qwandor reviewed (commented):
Weird! Good luck. I haven't finished reviewing this yet, but some comments in the meantime...

Serhii-the-Dev committed:
Move example-specific crates to dev dependencies.

Serhii-the-Dev committed:
Add watchdog timer to the Embassy example.

Serhii-the-Dev commented:
@qwandor Thank you for pointing those out, I've fixed the issues you mentioned. Also I realized that my MCUs are restarting due to the free watchdog timer timeout. According to the GD32F1x0 user manual, the watchdog timer is always running which makes me confused why the examples were running before without chip reset.
I've added a watchdog timer loop into the `embassy` example however it's still looking odd.

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+    callback: Cell<Option<(fn(*mut ()), *mut ())>>,
+}
+
+unsafe impl Send for AlarmState {}
```
If this is sound then please add a safety comment explaining how `AlarmState` is guaranteed to meet the safety requirements of `Send`.

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+    ((period as u64) << 15) + ((counter ^ ((period & 1) << 15)) as u64)
+}
+
+const ALARM_STATE_NEW: AlarmState = AlarmState::new();
```
Why this constant, rather than just calling `AlarmState::new()` where you need it? Having both seems like unneccesary duplication.

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    fn get_alarm<'a>(&'a self, cs: CriticalSection<'a>, alarm: AlarmHandle) -> &'a AlarmState {
+        // safety: we're allowed to assume the AlarmState is created by us, and
```
Do you mean the `AlarmHandle`? This is a safe method, so you can't assume anything about what arguments it is called with. Either make this method `unsafe` and document the safety requirement, or use a safe method below to panic if the handle is out of bounds rather than causing undefined behaviour.

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+        let period = self.period.load(Ordering::Relaxed);
+        compiler_fence(Ordering::Acquire);
+        // Timer1 has a 32-bit counter, while other timers resolution is limited to 16 bits
+        let counter = r.cnt().read().cnt().bits() as u32;
```
Can you use `.into()` here rather than `as u32`? That way it's clear that no bits are being discarded.

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+    fn now(&self) -> u64 {
+        let r = timer();
+        let period = self.period.load(Ordering::Relaxed);
+        compiler_fence(Ordering::Acquire);
```
Would using `Ordering::Acquire` for the period load operation on the previous line have an equivalent effect?

qwandor commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    unsafe fn allocate_alarm(&self) -> Option<AlarmHandle> {
+        critical_section::with(|_| {
```
Why does this need a critical section, when it's only accessing atomics?

Serhii-the-Dev committed:
bump to the latest Embassy version

qwandor committed:
Store timer in EmbassyTimeDriver rather than getting it unsafely.

Cleaned up some other minor issues too.

qwandor committed:
Fix ordering of atomic operations.

Serhii-the-Dev commented on src/embassy/time_driver.rs:
```diff
@@ -0,0 +1,363 @@
+
+    fn on_interrupt(&self) {
+        critical_section::with(|cs| {
+            let timer = self.timer.try_get().unwrap().borrow(cs);
```
@qwandor I now have a crash in runtime here:
```
panicked at /home/serhii/work/gd32f1x0-hal/src/embassy/time_driver.rs:214:46:
called `Option::unwrap()` on a `None` value
Firmware exited unexpectedly: Multiple
Core 0
    Frame 0: HardFault_ @ 0x08005410
    ...
```

qwandor replied:
I guess that the interrupt was happening before `EmbassyTimeDriver::init` called `DRIVER.timer.init`. I've moved the interrupt unmasking after that call, see if it works now.

Serhii-the-Dev replied:
Indeed it was the reason. The example is now operational as well as my GD32 project. Can't say if the timer is super-accurate though it works fine enough to provide a software UART capabilities.

qwandor replied:
Thanks for testing!

qwandor committed:
Unmask interrupts after initialising timer, to avoid race condition.

qwandor commented:
Let me know if you want a new release with this, or have other PRs to go first.
